File: scripts/check-real-agent-workflow.py
#!/usr/bin/env python3
"""Check the real-agent workflow's runtime and security contract."""

import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
WORKFLOW = ROOT / ".github" / "workflows" / "real-agent-e2e.yml"
HELPER = ROOT / "scripts" / "real-agent-e2e.py"
PINS = {
    "actions/checkout": "11bd71901bbe5b1630ceea73d27597364c9af683",
    "actions/upload-artifact": "ea165f8d65b6e75b540449e92b4886f43607fa02",
}


def check():
    text = WORKFLOW.read_text(encoding="utf-8")
    helper = HELPER.read_text(encoding="utf-8")
    contract = text + "\n" + helper
    assert "workflow_call:" in text and "workflow_dispatch:" in text
    assert "permissions: {}" in text
    assert "repository:" in text and "expected_sha:" in text
    assert "persist-credentials: false" in text
    assert "@11bd71901bbe5b1630ceea73d27597364c9af683" in text
    assert "@ea165f8d65b6e75b540449e92b4886f43607fa02" in text
    assert "@0.148.0" in text and "@v" not in text.split("@0.148.0", 1)[0]
    for required in (
        "\"codex\", \"exec\"", "--json", "--ephemeral", "--ignore-user-config",
        "--sandbox", "--ask-for-approval", "OPENAI_API_KEY", "AGENT_GITHUB_TOKEN",
        "network_proxy", "if: always()", "retention-days: 7",
    ):
        assert required in contract, required
    assert "--dangerously-bypass-approvals-and-sandbox" not in text
    assert "BOOTSTRAP_E2E_TOKEN" not in text
    assert "status:approved" not in text
    uses = re.findall(r"^\s*uses:\s*([^\s#]+)", text, re.MULTILINE)
    for reference in uses:
        action, separator, sha = reference.partition("@")
        assert separator and re.fullmatch(r"[0-9a-f]{40}", sha), reference
        if action in PINS:
            assert PINS[action] == sha, reference
    print("real-agent workflow static check OK")


if __name__ == "__main__":
    check()

File: scripts/typed-runtime/check-real-agent-workflow.js
#!/usr/bin/env node
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const pins = { "actions/checkout": "11bd71901bbe5b1630ceea73d27597364c9af683", "actions/upload-artifact": "ea165f8d65b6e75b540449e92b4886f43607fa02" };
export const check = async (projectRoot = root) => {
    const workflow = await readFile(path.join(projectRoot, ".github/workflows/real-agent-e2e.yml"), "utf8");
    const helper = await readFile(path.join(projectRoot, "scripts/real-agent-e2e.mjs"), "utf8");
    const journeyWorkflow = await readFile(path.join(projectRoot, ".github/workflows/real-agent-journey.yml"), "utf8");
    const journeyReporter = await readFile(path.join(projectRoot, "scripts/real-agent-journey.mjs"), "utf8");
    const contract = `${workflow}\n${helper}`;
    const required = ["workflow_call:", "workflow_dispatch:", "permissions: {}", "repository:", "expected_sha:", "persist-credentials: false", "@0.148.0", 'spawnSync("codex", ["exec"', "--json", "--ephemeral", "--ignore-user-config", "--sandbox", "--ask-for-approval", "OPENAI_API_KEY", "AGENT_GITHUB_TOKEN", "network_proxy", "if: always()", "retention-days: 7"];
    for (const value of required)
        if (!contract.includes(value))
            throw new Error(`real-agent workflow is missing ${value}`);
    if (workflow.includes("python") || workflow.includes("--dangerously-bypass-approvals-and-sandbox") || workflow.includes("BOOTSTRAP_E2E_TOKEN") || workflow.includes("status:approved"))
        throw new Error("real-agent workflow contains a retired or unsafe contract");
    if (!journeyWorkflow.includes("node scripts/real-agent-journey.mjs report") || !journeyWorkflow.includes("--input evidence/journey.json") || !journeyWorkflow.includes('--artifact-url "$WORKFLOW_URL"') || !journeyReporter.includes('command === "report"') || !journeyReporter.includes("canonicalIssue"))
        throw new Error("real-agent journey report command is missing or incomplete");
    for (const reference of [...workflow.matchAll(/^\s*uses:\s*([^\s#]+)/gmu)].map((match) => match[1])) {
        const [action, sha] = reference.split("@");
        if (!/^[0-9a-f]{40}$/u.test(sha ?? ""))
            throw new Error(`action is not pinned: ${reference}`);
        if (pins[action] && pins[action] !== sha)
            throw new Error(`action pin is not verified: ${reference}`);
    }
    return "real-agent workflow static check OK";
};
if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href)
    check().then((message) => console.log(message)).catch((error) => { console.error(error.message); process.exitCode = 1; });
